zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/Kubectl.java

```diff
@@ -0,0 +1,157 @@
+    return execAndGet(new String[]{"delete", "-f", "-"}, spec);
+  }
+
+  public String wait(String resource, String waitFor, int timeoutSec) throws IOException {
+    try {
+      return execAndGet(new String[]{
+          "wait",
+          resource,
+          String.format("--for=%s", waitFor),
+          String.format("--timeout=%ds", timeoutSec)});
+    } catch (IOException e) {
+      if ("delete".equals(waitFor) && e.getMessage().contains("NotFound")) {
+        LOGGER.info("{} Not found. Maybe already deleted.", resource);
+        return "";
+      } else {
+        throw e;
+      }
+    }
+  }
+
+  public ExecuteWatchdog portForward(String resource, String [] ports) throws IOException {
+    DefaultExecutor executor = new DefaultExecutor();
+    CommandLine cmd = new CommandLine(kubectlCmd);
+    cmd.addArguments("port-forward");
+    cmd.addArguments(resource);
+    cmd.addArguments(ports);
+
+    ExecuteWatchdog watchdog = new ExecuteWatchdog(-1);
+    executor.setWatchdog(watchdog);
+
+    executor.execute(cmd, new ExecuteResultHandler() {
+      @Override
+      public void onProcessComplete(int i) {
+        LOGGER.info("Port-forward stopped");
+      }
+
+      @Override
+      public void onProcessFailed(ExecuteException e) {
+        LOGGER.debug("port-forward process exit", e);
+      }
+    });
+
+    return watchdog;
+  }
+
+  String execAndGet(String [] args) throws IOException {
+    return execAndGet(args, "");
+  }
+
+  @VisibleForTesting
+  String execAndGet(String [] args, String stdin) throws IOException {
+    InputStream ins = IOUtils.toInputStream(stdin);
+    ByteArrayOutputStream stdout = new ByteArrayOutputStream();
+    ByteArrayOutputStream stderr = new ByteArrayOutputStream();
+    ArrayList<String> argsToOverride = new ArrayList<>(Arrays.asList(args));
+
+    // set namespace
+    if (namespace != null) {
+      argsToOverride.add("--namespace=" + namespace);
+    }
+
+    LOGGER.info("kubectl " + argsToOverride);
+    LOGGER.debug(stdin);
+
+    try {
+      int exitCode = execute(
+              argsToOverride.toArray(new String[0]),
+              ins,
+              stdout,
+              stderr
+      );
+
+      if (exitCode == 0) {
+        String output = new String(stdout.toByteArray());
+        return output;
+      } else {
+        String output = new String(stderr.toByteArray());
+        throw new IOException(String.format("non zero return code (%d). %s", exitCode, output));
+      }
+    } catch (Exception e) {
+      String output = new String(stderr.toByteArray());
+      throw new IOException(output, e);
+    }
+  }
+
+  public int execute(String [] args, InputStream stdin, OutputStream stdout, OutputStream stderr) throws IOException {
+    DefaultExecutor executor = new DefaultExecutor();
+    CommandLine cmd = new CommandLine(kubectlCmd);
+    cmd.addArguments(args);
+
+    ExecuteWatchdog watchdog = new ExecuteWatchdog(60 * 1000);
```

Comment:
@Leemoonsoo This value should be the same as the kubectl timeout flags. In the case of kubectl wait: https://github.com/apache/zeppelin/pull/3240/files#diff-7b231819fa2236eab55c934add3b56e8R69

This timeout may be reached before kubectl wait's timeout, causing an exception to be thrown:
```
 INFO [2019-05-03 03:24:59,232] ({SchedulerFactory3} Kubectl.java[execAndGet]:121) - kubectl [wait, pod/spark-foqspk, --for=condition=Ready, --timeout=600s, --namespace=aikwei-sng]
 WARN [2019-05-03 03:25:59,257] ({SchedulerFactory3} NotebookServer.java[onStatusChange]:1671) - Job paragraph_1556788652522_1837703941 is finished, status: ERROR, exception: null, result: %text org.apache.zeppelin.interpreter.InterpreterException: java.io.IOException: 
	at org.apache.zeppelin.interpreter.remote.RemoteInterpreter.open(RemoteInterpreter.java:134)
	at org.apache.zeppelin.interpreter.remote.RemoteInterpreter.getFormType(RemoteInterpreter.java:298)
	at org.apache.zeppelin.notebook.Paragraph.jobRun(Paragraph.java:417)
	at org.apache.zeppelin.notebook.Paragraph.jobRun(Paragraph.java:76)
	at org.apache.zeppelin.scheduler.Job.run(Job.java:172)
	at org.apache.zeppelin.scheduler.AbstractScheduler.runJob(AbstractScheduler.java:121)
	at org.apache.zeppelin.scheduler.RemoteScheduler$JobRunner.run(RemoteScheduler.java:187)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: 
	at org.apache.zeppelin.interpreter.launcher.Kubectl.execAndGet(Kubectl.java:141)
	at org.apache.zeppelin.interpreter.launcher.Kubectl.execAndGet(Kubectl.java:106)
	at org.apache.zeppelin.interpreter.launcher.Kubectl.wait(Kubectl.java:65)
	at org.apache.zeppelin.interpreter.launcher.K8sRemoteInterpreterProcess.start(K8sRemoteInterpreterProcess.java:92)
	at org.apache.zeppelin.interpreter.ManagedInterpreterGroup.getOrCreateInterpreterProcess(ManagedInterpreterGroup.java:65)
	at org.apache.zeppelin.interpreter.remote.RemoteInterpreter.getOrCreateInterpreterProcess(RemoteInterpreter.java:110)
	at org.apache.zeppelin.interpreter.remote.RemoteInterpreter.internal_create(RemoteInterpreter.java:163)
	at org.apache.zeppelin.interpreter.remote.RemoteInterpreter.open(RemoteInterpreter.java:131)
	... 13 more
Caused by: org.apache.commons.exec.ExecuteException: Process exited with an error: 143 (Exit value: 143)
	at org.apache.commons.exec.DefaultExecutor.executeInternal(DefaultExecutor.java:404)
	at org.apache.commons.exec.DefaultExecutor.execute(DefaultExecutor.java:166)
	at org.apache.commons.exec.DefaultExecutor.execute(DefaultExecutor.java:153)
	at org.apache.zeppelin.interpreter.launcher.Kubectl.execute(Kubectl.java:155)
	at org.apache.zeppelin.interpreter.launcher.Kubectl.execAndGet(Kubectl.java:125)
	... 20 more
```

Reply:
Thanks for pointing it out!